doc: add plugins contribution doc (#4732)
operator-framework/operator-sdk: website/content/en/docs/contribution-guidelines/plugins.md

All Language-based operator projects provided by SDK (Anisble/Helm/Golang) follows its standard and have a common base. By following its standard, the common base is generate using [kustomize][kustomize] which is a project maintained and adopted by Kubernetes community to help work with the manifests (`ymal` files) that are used to configure the resources on the clusters. 

Then, the specific files for each language are scaffolded by its plugins. In this way, SDK CLI tool is able to provide custom plugins and helpers which are common for all language-types. To check the common base you can also run:

```sh
operator-sdk init --plugins=kustomize
```

Also, see the topic [Language-based Plugins][kb-language-plugins] to understand know how it works.

### Common scaffolds

| File/Directory | Description | 
| ------ | ----- |
|  Dockerfile |  Defines the operator(manager) image |  
|  Makefile |  Provide the helpers and options for the users. (e.g. `make bundle` which generate/update the OLM [bundle][bundle] manifests ) |  
|  PROJECT |  Project configuration. Store the data used to do the scaffolds. For further information see [Project Config][kb-project] | 

> |  bundle.Dockerfile |  Docker image which is used to provide the helpers to integate the project with OLM. (e.g. [operator-sdk run ./bundle][sdk-cli-run-bundle]) | 

Reviewer: ```suggestion
|  bundle.Dockerfile | Docker image which is used to provide the helpers to integrate the project with OLM. (e.g. [operator-sdk run ./bundle][sdk-cli-run-bundle]) | 
```

|  config/ |  Directory which has all [kustomize's][kustomize] manifest to configure and test the project | 

You can check the [Project Layout][project-layout] to understand better the files and directories which are scaffold by SDK CLI which are common or not for each Language-type. 

## Custom Plugins

By default, some plugins are used by SDK to add the required changes in the scaffolds to provide its features which currently are: 

> - [manifests.sdk.operatorframework.io][plugin-manifest]: resposanble to perform all required changes in the scaffolds and provide the helpers to allow the projects to be integrated with OLM. 

Reviewer: ```suggestion
- [manifests.sdk.operatorframework.io][plugin-manifest]: responsible to perform all required changes in the scaffolds and provide the helpers to allow the projects to be integrated with OLM. 
```

> - [scorecard.sdk.operatorframework.io][plugin-scorecard]: resposanble to scaffold all files and helpers required to provide the [Scorecard][scorecard] feature

Reviewer: ```suggestion
- [scorecard.sdk.operatorframework.io][plugin-scorecard]: responsible to scaffold all files and helpers required to provide the [Scorecard][scorecard] feature
```

[bundle]: https://github.com/operator-framework/operator-registry/blob/v1.16.1/docs/design/operator-bundle.md#operator-bundle
[kb-project]: https://deploy-preview-2115--kubebuilder.netlify.app/plugins/project-config.html
[sdk-cli-run-bundle]: ../../docs/cli/operator-sdk_run
[project-layout]: ../../docs/overview/project-layout
[plugin-manifest]: https://github.com/operator-framework/operator-sdk/tree/master/internal/plugins/manifests/v2
[plugin-scorecard]: https://github.com/operator-framework/operator-sdk/tree/master/internal/plugins/scorecard/v2
[kubebuilder-declarative-pattern]: https://github.com/kubernetes-sigs/kubebuilder-declarative-pattern
[kubebuilder-declarative-pattern-example]: https://github.com/kubernetes-sigs/kubebuilder/blob/master/testdata/project-v3/controllers/firstmate_controller.go
[default-scaffold]: https://github.com/kubernetes-sigs/kubebuilder/blob/master/testdata/project-v3/controllers/admiral_controller.go
[kb-issue]: https://github.com/kubernetes-sigs/kubebuilder/issues/1378
[create-your-own-plugins]: https://deploy-preview-2115--kubebuilder.netlify.app/plugins/creating-plugins.html
[scorecard]: ../../docs/advanced-topics/scorecard

> [kubebuilder]: https://github.com/kubernetes-sigs/kubebuilder 

Author: The `https://deploy-preview-2115--kubebuilder` requires to be replaced when the PR https://github.com/kubernetes-sigs/kubebuilder/pull/2115 gt merged.

Then, the specific files for each language are scaffolded for the language's plugins. In this way, the SDK CLI tool can provide custom plugins and helpers which are common and useful for all language types. To check the common base, you can also run:

```sh
operator-sdk init --plugins=kustomize
```

Also, see the topic [Language-based Plugins][kb-language-plugins] to understand how it works.

### Common scaffolds 

Following the default common scaffolds for the projects which are built with SDK.

| File/Directory | Description | 
| ------ | ----- |
|  Dockerfile |  Defines the operator(manager) image |  
|  Makefile |  Provides the helpers and options for the users. (e.g. `make bundle` which generate/update the OLM [bundle][bundle] manifests ) |  
|  PROJECT |  Project configuration. Stores the data used to do the scaffolds. For further information see [Project Config][kb-project] | 
|  bundle.Dockerfile | Docker image which is used to provide the helpers to integrate the project with OLM. (e.g. [operator-sdk run ./bundle][sdk-cli-run-bundle]) | 
|  config/ |  Directory which has all [kustomize's][kustomize] manifest to configure and test the project | 

> You can check the [Project Layout][project-layout] to better understand the files and directories scaffold by SDK CLI, which are common or not for each Language-type. 

Reviewer 2: Nit: 
```suggestion
You can check the [Project Layout][project-layout] to better understand the files and directories scaffolded by SDK CLI, which may be common for each language-type. 
```

## Custom Plugins

> By default, some plugins are used by SDK to add the required changes in the scaffolds to provide its features which currently are: 

Reviewer 3: By default plugins are used by the Operator SDK to provide the following features:

- [manifests.sdk.operatorframework.io][plugin-manifest]: perform the required scaffolds to provide the helpers to allow the projects to be integrated with OLM. 
- [scorecard.sdk.operatorframework.io][plugin-scorecard]: perform the required scaffolds to provide the [Scorecard][scorecard] feature.

### Optional/custom plugins

> Users also can use custom plugins when they execute the SDK CLI sub-commands as a helper for its options which would like:

Reviewer 2: Nit:
```suggestion
Users can also use custom plugins when they execute the SDK CLI sub-commands as a helper for its options which could be like:
```

> Users can also use custom plugins when they execute the SDK CLI sub-commands as a helper for its options which could be like:

Reviewer 3: Users can also use custom plugins when they execute the SDK CLI sub-commands as a helper which includes the following options:

```sh
operator-sdk create api --plugins="go/v3,declarative"
```

> The above example will scaffold a custom code in the controllers after an API is created to allow the users to develop solutions using the [kubebuilder declarative pattern][kubebuilder-declarative-pattern]. (e.g [default scaffold][default-scaffold] versus [example][kubebuilder-declarative-pattern-example]).

Reviewer 3: The above example will scaffold custom code in the controllers after an API is created to allow the users to develop solutions using the [kubebuilder declarative pattern][kubebuilder-declarative-pattern]. (e.g [default scaffold][default-scaffold] versus [example][kubebuilder-declarative-pattern-example]).

> The above example will scaffold a custom code in the controllers after when an api is created to allow the users develop solutions using the [kubebuilder declarative pattern][kubebuilder-declarative-pattern]. (e.g [default scaffold][default-scaffold] versus [example][kubebuilder-declarative-pattern-example]).

Reviewer 3: The above example will scaffold a custom code in the controllers after an api is created to allow the users to develop solutions using the [kubebuilder declarative pattern][kubebuilder-declarative-pattern]. (e.g [default scaffold][default-scaffold] versus [example][kubebuilder-declarative-pattern-example]).

Author: done.

> Note that custom plugins which are called via the init sub-command, will be added in the `layout` field of the PROJECT file and will work such as global plugins. Then, it means that for any sub-command executed; they will also be called. 

Reviewer 3: Note that custom plugins are called via the init sub-command to work as global plugins, and will be added in the `layout` field of the PROJECT file. Any sub-command executed will then also be called.

Author: done.

## Plugins Vision

> Maintainers and contributors are able to create their own plugin(s) using the same standards and approaches. Theoretically, that could be an option to allow these solutions to be compatible and useful with SDK CLI in the future or at least could result in lower effort to add them on this project, if that is its ultimate motivation. Currently, SDK cannot discover and use plugins that are not ship or build with its binary. However, it has been discussed in the issue [kubebuilder/issues/1378][kb-issue].

Reviewer 2: ```suggestion
Maintainers and contributors are able to create their own plugin(s) using the same standards and approaches. Theoretically, that could be an option to allow these solutions to be compatible and useful with SDK CLI in the future or at least could result in lower effort to add them on this project, if that is its ultimate motivation. Currently, SDK cannot discover and use plugins that are not shipped or built with its binary. However, it has been discussed in the issue [kubebuilder/issues/1378][kb-issue].
```

## How to create your own plugins

> If you are looking to develop similar solutions to allow users, for example, to create projects using other languages or that could work as helpers for the projects built with SDK/Kubebuilder projects, then see the [Creating your own plugins][create-your-own-plugins] to know how you can do that following this approach as its benefits.

Reviewer 3: If you are looking to develop similar solutions to allow users for example to create projects using other languages or that could work as helpers for the projects built with SDK/Kubebuilder projects, then see the [Creating your own plugins][create-your-own-plugins] to see how you can benefit and apply this approach.
